File: exercises/06_lifecycle/solution/payments/temporal/review_starter.py
import asyncio

from temporalio.client import Client

from compliance.domain import ComplianceResult
from payments.domain import TASK_QUEUE
from payments.temporal.workflows import ReviewCallerWorkflow
from shared.domain import ReviewRequest

NAMESPACE = "payments-namespace"


async def main() -> None:
    print("==========================================================")
    print("  REVIEW STARTER - Submitting review for TXN-B via Nexus")
    print("==========================================================\n")

    client = await Client.connect("localhost:7233", namespace=NAMESPACE)

    # Change approved to False to deny instead
    request = ReviewRequest(
        transaction_id="TXN-B",
        approved=True,
        explanation="Approved after manual review",
    )

    print("  Submitting review for TXN-B via Nexus...")
    print(f"  Approved: {request.approved}")
    print(f"  Explanation: {request.explanation}")
    print()

    result: ComplianceResult = await client.execute_workflow(
        ReviewCallerWorkflow.submit_review,
        request,
        id="review-TXN-B",
        task_queue=TASK_QUEUE,
    )

    approved_str = "APPROVED" if result.approved else "DENIED"
    print(f"  Review result: {approved_str}")
    print(f"  Risk level:    {result.risk_level}")
    print(f"  Explanation:   {result.explanation}")
    print()
    print("  TXN-B review submitted! The payment workflow will now complete.")
    print("==========================================================")


if __name__ == "__main__":
    asyncio.run(main())

File: polyglot/java-legacy/src/main/java/payments/temporal/ReviewStarter.java
package payments.temporal;

import compliance.domain.ComplianceResult;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowClientOptions;
import io.temporal.client.WorkflowOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;
import payments.Shared;
import shared.domain.ReviewRequest;

/**
 * [GIVEN] Starts a ReviewCallerWorkflow to approve TXN-B via Nexus.
 *
 * Instead of running 'temporal workflow update execute' directly, this starter
 * triggers a ReviewCallerWorkflow that calls the submitReview Nexus operation.
 * The Compliance team's sync handler forwards the decision to the running
 * ComplianceWorkflow as a Workflow Update.
 *
 * Change 'approved' to false below to see the denial path.
 */
public class ReviewStarter {

    public static void main(String[] args) {
        System.out.println("==========================================================");
        System.out.println("  REVIEW STARTER — Submitting review for TXN-B via Nexus");
        System.out.println("==========================================================\n");

        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClientOptions clientOptions = WorkflowClientOptions.newBuilder()
                .setNamespace("payments-namespace")
                .build();
        WorkflowClient client = WorkflowClient.newInstance(service, clientOptions);

        // Change approved to false to deny instead
        ReviewRequest request = new ReviewRequest("TXN-B", true, "Approved after manual review");

        ReviewCallerWorkflow workflow = client.newWorkflowStub(
                ReviewCallerWorkflow.class,
                WorkflowOptions.newBuilder()
                        .setTaskQueue(Shared.TASK_QUEUE)
                        .setWorkflowId("review-TXN-B")
                        .build());

        System.out.println("  Submitting review for TXN-B via Nexus...");
        System.out.println("  Approved: " + request.isApproved());
        System.out.println("  Explanation: " + request.getExplanation());
        System.out.println();

        ComplianceResult result = workflow.submitReview(request);

        System.out.println("  Review result: " + (result.isApproved() ? "✅ APPROVED" : "🚫 DENIED"));
        System.out.println("  Risk level:    " + result.getRiskLevel());
        System.out.println("  Explanation:   " + result.getExplanation());
        System.out.println();
        System.out.println("  TXN-B review submitted! The payment workflow will now complete.");
        System.out.println("==========================================================");

        System.exit(0);
    }
}
